### PR TITLE
Resolve #17 optimize use of '^' / '\A'

### LINKED ISSUE #17: optimize use of `^`/`\A`
At the moment, the VM always scans the [entire input](https://github.com/rust-lang/regex/blob/master/src/vm.rs#L139) unless one of the two following things happens:

* The regex has a *literal prefix* and that prefix cannot be found anywhere in the input. (This always scans the entire input.)
* A match is found. This may *not* scan the entire input, but it requires a match.

Ideally, given a regex `^{something}...`, it could terminate the search early if `{something}` doesn't match at the start of the input.

### PR DESCRIPTION
**Pull request by tmerr**
The problem was that a regular expression like `^x` would take linear time to is_match a string `aaa....` when all it needed to do was check the first character then bail out. This PR makes it so if the regex begins with `^` (without multiline) or `\A`, then it stops iterating over characters as soon as there are no possible matches.

Benchmarks:

```
dynamic:
anchored_literal_long_non_match: 7492 ns/iter => 435 ns/iter
anchored_literal_short_non_match: 917 ns/iter => 431 ns/iter

native:
anchored_literal_long_non_match: 5020 ns/iter => 50 ns/iter
anchored_literal_short_non_match: 375 ns/iter => 50 ns/iter
```

### CODE REVIEW

**Comment by huonw** [comment]
Thanks!

Could I ask you to expand the commit message a little to describe what the problem is and how it is fixed? Just saying the issue number is better than nothing, but it does mean that someone looking the source and checking the git history needs to cross reference against github (which may have disappeared at that point).

(I'm happy to help if you need assistance with git or anything else.)

**Comment by tmerr** [comment]
Sure, expanded. Is that alright?

File: src/vm.rs [hunk 1]
```diff
@@ -144,12 +144,18 @@ impl<'r, 't> Nfa<'r, 't> {
...
+                if self.ic != 0 && prefix_anchor {
+                    break;
+                }
+                
                 // If there are no threads to try, then we'll have to start
                 // over at the beginning of the regex.
                 // BUT, if there's a literal prefix for the program, try to
                 // jump ahead quickly. If it can't be found, then we can bail
                 // out early.
-                if self.prog.prefix.len() > 0 && clist.size == 0 {
```
Comment (BurntSushi) [inline src/vm.rs]: Why did you remove `&& clist.size == 0`? That doesn't seem right to me. I think literal prefixes should only be tried if there are no other threads to try. Moreover, that check is still in place in the `regex!` macro code: https://github.com/tmerr/regex/blob/master/regex_macros/src/lib.rs#L548-L553

Comment (BurntSushi) [inline src/vm.rs]: I'm kind of concerned that this doesn't cause a test to fail... Perhaps I need to come up with one. :-)

**Comment by BurntSushi** [comment]
@tmerr This looks great. I'm happy to see that the change is small. I just had one question, but otherwise this looks awesome. Thank you!

File: src/vm.rs (hunk 1 above)
Comment (huonw) [inline src/vm.rs]: This whole section of code seems to be inside a `if clist.size == 0 {` branch.

Comment (BurntSushi) [inline src/vm.rs]: Ah... doy. Thanks :-)

Comment (tmerr) [inline src/vm.rs]: Ahh it's hard to see from the diff snippets but in the original code it was redundant
```
if clist.size == 0 {
    ...
    if self.prog.prefix.len() > 0 && clist.size == 0 {
    ...
    }
}
```
It's not my top priority, I just tried to sneak it the removal :P

**Comment by huonw** [comment]
@tmerr that messsage looks much better, thanks! Could you place it into the git commit message (e.g. `git commit --amend` and then `git push -f` to update the PR by force pushing over this branch.)

File: src/vm.rs (hunk 1 above)
Comment (tmerr) [inline src/vm.rs]: I'm perfectly happy to revert that part if you want

Comment (BurntSushi) [inline src/vm.rs]: Oh no no. It's great. :-) It's my misunderstanding!

**Comment by tmerr** [comment]
@huonw Done, thanks for the help!

It's been a blast reading through this code by the way. I had been aware of NFAs but only  the "draw pretty circles on paper with lines between them" version, not as a machine with split and jump instructions. The two models of thinking about it meld together nicely

**Comment by huonw** [comment]
Thanks! I merged as 10edff6bb76f1c9f4423f3bb8ad373d0c7a69773 (there was a piece of trailing whitespace which I took the initiative to remove).
